paychmgr/manager.go

```diff
@@ -0,0 +1,278 @@
+package paychmgr
+
+import (
+	"context"
+	"sync"
+
+	"github.com/ipfs/go-datastore"
+
+	"golang.org/x/sync/errgroup"
+
+	xerrors "golang.org/x/xerrors"
+
+	"github.com/filecoin-project/lotus/api"
+
+	"github.com/filecoin-project/specs-actors/actors/builtin/paych"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log/v2"
+	"go.uber.org/fx"
+
+	"github.com/filecoin-project/go-address"
+
+	"github.com/filecoin-project/lotus/chain/stmgr"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/node/impl/full"
+)
+
+var log = logging.Logger("paych")
+
+type ManagerApi struct {
+	fx.In
+
+	full.MpoolAPI
+	full.WalletAPI
+	full.StateAPI
+}
+
+type StateManagerApi interface {
+	LoadActorState(ctx context.Context, a address.Address, out interface{}, ts *types.TipSet) (*types.Actor, error)
+	Call(ctx context.Context, msg *types.Message, ts *types.TipSet) (*api.InvocResult, error)
+}
+
+type Manager struct {
+	// The Manager context is used to terminate wait operations on shutdown
+	ctx      context.Context
+	shutdown context.CancelFunc
+
+	store  *Store
+	sm     StateManagerApi
+	sa     *stateAccessor
+	pchapi paychApi
+
+	lk       sync.RWMutex
+	channels map[string]*channelAccessor
+
+	mpool  full.MpoolAPI
+	wallet full.WalletAPI
+	state  full.StateAPI
+}
+
+type paychAPIImpl struct {
+	full.MpoolAPI
+	full.StateAPI
+}
+
+func NewManager(sm *stmgr.StateManager, pchstore *Store, api ManagerApi) *Manager {
+	return &Manager{
+		store:    pchstore,
+		sm:       sm,
+		sa:       &stateAccessor{sm: sm},
+		channels: make(map[string]*channelAccessor),
+		// TODO: Is this the correct way to do this or can I do something different
+		// with dependency injection?
+		pchapi: &paychAPIImpl{api.MpoolAPI, api.StateAPI},
+
+		mpool:  api.MpoolAPI,
+		wallet: api.WalletAPI,
+		state:  api.StateAPI,
+	}
+}
+
+// newManager is used by the tests to supply mocks
+func newManager(sm StateManagerApi, pchstore *Store, pchapi paychApi) (*Manager, error) {
+	pm := &Manager{
+		store:    pchstore,
+		sm:       sm,
+		sa:       &stateAccessor{sm: sm},
+		channels: make(map[string]*channelAccessor),
+		pchapi:   pchapi,
+	}
+	return pm, pm.Start(context.Background())
+}
+
+// HandleManager is called by dependency injection to set up hooks
+func HandleManager(lc fx.Lifecycle, pm *Manager) {
+	lc.Append(fx.Hook{
+		OnStart: func(ctx context.Context) error {
+			return pm.Start(ctx)
+		},
+		OnStop: func(context.Context) error {
+			return pm.Stop()
+		},
+	})
+}
+
+// Start checks the datastore to see if there are any channels that have
+// outstanding add funds messages, and if so, waits on the messages.
+// Outstanding messages can occur if an add funds message was sent
+// and then lotus was shut down or crashed before the result was
+// received.
+func (pm *Manager) Start(ctx context.Context) error {
+	pm.ctx, pm.shutdown = context.WithCancel(ctx)
```

Comment:
Is this correct - should we shutdown when the context passed to Start() is cancelled?

Reply:
If this is passed to `fx.Lifecycle`, this context is only alive for the startup process (~5s), you probably want to construct a context in the constructor from `mctx helpers.MetricsCtx, lc fx.Lifecycle`, using `ctx := helpers.LifecycleCtx(mctx, lc)`